File: web/src/services/tracker/TrackerItemService.js
import { request, METHOD } from "@/utils/request";
export function findTrackerItems(
  projectId,
  trackerId,
  sprintId,
  filter,
  keyword,
  pageable,
  sort,
  isTest
) {
  let config = {};
  return request({
    url:
      "/tracker/project/items/list?" + sort?.map((_) => `sort=${_}`).join("&"),
    method: METHOD.POST,
    params: {
      projectId: projectId,
      trackerId: trackerId,
      sprintId: sprintId,
      keyword: keyword,
      ...pageable,
      isTest,
    },
    data: filter,
  });
}

export function findTrackerItemsTree(
  projectId,
  trackerId,
  sprintId,
  filter,
  keyword,
  pageable,
  sort,
  linkDire,
  linkType,
  linkLevel,
) {
  let config = {};
  return request({
    url:
      "/tracker/project/items/list-tree?" + sort?.map((_) => `sort=${_}`).join("&"),
    method: METHOD.POST,
    params: {
      projectId: projectId,
      trackerId: trackerId,
      sprintId: sprintId,
      keyword: keyword,
      ...pageable,
      linkDire,
      linkType,
      linkLevel
    },
    data: filter,
  });
}


export function findGroupItems(
  projectId,
  trackerId,
  sprintId,
  filter,
  keyword
) {
  let config = {};
  return request({
    url: "/tracker/project/items/group",
    method: METHOD.POST,
    params: {
      projectId: projectId,
      trackerId: trackerId,
      sprintId: sprintId,
      keyword: keyword,
    },
    data: filter,
  });
}

export function exportTrackerItems(
  projectId,
  trackerId,
  sprintId,
  filter,
  selectedRowIds,
  pageable,
  sort
) {
  return request({
    url:
      "/tracker/project/items/export?" +
      sort.map((_) => `sort=${_}`).join("&") +
      "&" +
      selectedRowIds.map((_) => `selectedRowIds=${_}`).join("&"),
    method: METHOD.POST,
    params: {
      projectId: projectId,
      trackerId: trackerId,
      sprintId: sprintId,
      ...pageable,
    },
    data: filter,
    responseType: "blob",
  });
}
export function downloadImportDome(trackerId) {
  return request({
    url: `/tracker/project/items/import-demo`,
    method: METHOD.GET,
    params: {
      trackerId,
    },
    responseType: "blob",
  });
}
export function importTrackerItems(trackerId, uploadedFile) {
  return request({
    url: `/tracker/project/items/import`,
    method: METHOD.GET,
    params: {
      trackerId,
      uploadedFile,
    },
  });
}
export function findOneTrackerItem(itemId) {
  return request({
    url: `/tracker/project/items/${itemId}`,
    method: METHOD.GET,
  });
}
export function findTrackerItemByIds(itemIds) {
  return request({
    url:"/tracker/project/items/byIds",
    method: METHOD.POST,
    data: itemIds,
  });
}
export function createTrackerItem(parameter) {
  return request({
    url: "/tracker/project/items",
    method: METHOD.POST,
    data: parameter,
  });
}
export function batchUpdateTrackerItem(parameter) {
  return request({
    url: "/tracker/project/items/batch",
    method: METHOD.PUT,
    data: parameter,
  });
}

export function updateTrackerItem(parameter) {
  return request({
    url: "/tracker/project/items",
    method: METHOD.PUT,
    data: parameter,
  });
}
export function deleteTrackerItem(itemId) {
  return request({
    url: `/tracker/project/items/${itemId}`,
    method: METHOD.DELETE,
  });
}
export function deleteTrackerItemsByTrackerId(trackerId) {
  return request({
    url: `/tracker/project/items/deleteByTrackerId`,
    method: METHOD.DELETE,
    params: { trackerId: trackerId },
  });
}
export function batchDeleteTrackerItem(parameter) {
  return request({
    url: "/tracker/project/items/batch",
    method: METHOD.DELETE,
    data: parameter,
  });
}
export function createTrackerLink(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/link`,
    method: METHOD.POST,
    data: newVal,
  });
}
export function deleteTrackerLink(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/link`,
    method: METHOD.DELETE,
    data: newVal,
  });
}
export function findHyperlinkByObjectId(itemId) {
  return request({
    url: `/tracker/project/items/${itemId}/hyperlink`,
    method: METHOD.GET,
  });
}
export function updateHyperlink(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/hyperlink`,
    method: METHOD.PUT,
    data: newVal,
  });
}
export function deleteHyperlink(hyperlinkId) {
  return request({
    url: `/tracker/project/items/${hyperlinkId}/hyperlink`,
    method: METHOD.DELETE,
  });
}

export function updateTrackerLink(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/link`,
    method: METHOD.PUT,
    data: newVal,
  });
}
export function addWatch(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/add-watch`,
    method: METHOD.PUT,
    data: newVal,
  });
}
export function cancelWatch(itemId, newVal) {
  return request({
    url: `/tracker/project/items/${itemId}/cancel-watch`,
    method: METHOD.PUT,
    data: newVal,
  });
}

export function relatedWikis(itemId, newVal) {
  return request({
    url:
      `/tracker/project/items/${itemId}/relatedWikis?` +
      newVal.map((_) => `relatedWikiIds=${_}`).join("&"),
    method: METHOD.GET,
  });
}

export function changeSystemField(itemId, systemProperty, newValue) {
  return request({
    url: `/tracker/project/items/${itemId}/changeSystemField`,
    method: METHOD.POST,
    // params: { systemProperty: systemProperty},
    data: {systemProperty,newValue}
  });
}
export function changeCustomerField(itemId, fieldId, newValue) {
  return request({
    url: `/tracker/project/items/${itemId}/changeCustomerField`,
    method: METHOD.POST,
    // params: { fieldId: fieldId, newValue: newVal },
    data: {fieldId,newValue}
  });
}

export function stateChange(itemId, stateTransitionId) {
  return request({
    url: `/tracker/project/items/${itemId}/stateChange`,
    method: METHOD.POST,
    params: { stateTransitionId: stateTransitionId },
  });
}

export function findTrackerLogs(
  itemId,
  includeComment,
  includeChangeLog,
  includeAttachment,
  includeAssociation
) {
  return request({
    url: `/tracker/project/items/${itemId}/comments`,
    method: METHOD.GET,
    params: {
      includeComment: includeComment,
      includeChangeLog: includeChangeLog,
      includeAttachment: includeAttachment,
      includeAssociation: includeAssociation,
    },
  });
}
export function createComment(itemId, comment) {
  Object.assign(comment, { objectId: itemId, type: "COMMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/comments`,
    method: METHOD.POST,
    data: comment,
  });
}
export function saveComment(itemId, comment) {
  Object.assign(comment, { objectId: itemId, type: "COMMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/comments`,
    method: METHOD.PUT,
    data: comment,
  });
}
export function deleteComment(itemId, comment) {
  Object.assign(comment, { objectId: itemId, type: "COMMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/comments`,
    method: METHOD.DELETE,
    data: comment,
  });
}
export function replyComment(itemId, comment) {
  Object.assign(comment, { objectId: itemId, type: "COMMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/comments/reply`,
    method: METHOD.POST,
    data: comment,
  });
}

export function uploadAttachment(itemId, attachment) {
  Object.assign(attachment, { objectId: itemId, type: "ATTACHMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/attachments`,
    method: METHOD.POST,
    data: attachment,
  });
}
export function deleteAttachment(itemId, attachment) {
  Object.assign(attachment, { objectId: itemId, type: "ATTACHMENT" });
  return request({
    url: `/tracker/project/items/${itemId}/attachments`,
    method: METHOD.DELETE,
    data: attachment,
  });
}
export function findAttachmentsByObjectId(itemId) {
  return request({
    url: `/tracker/project/items/${itemId}/attachments`,
    method: METHOD.GET,
  });
}

export function findWorkHours(itemId) {
  return request({
    url: `/tracker/project/items/${itemId}/workHours`,
    method: METHOD.GET,
  });
}
export function createWorkHours(itemId, workHours) {
  Object.assign(workHours, { objectId: itemId, type: "WORKHOURS" });
  return request({
    url: `/tracker/project/items/${itemId}/workHours`,
    method: METHOD.POST,
    data: workHours,
  });
}
export function saveWorkHours(itemId, workHours) {
  Object.assign(workHours, { objectId: itemId, type: "WORKHOURS" });
  return request({
    url: `/tracker/project/items/${itemId}/workHours`,
    method: METHOD.PUT,
    data: workHours,
  });
}
export function deleteWorkHours(itemId, workHours) {
  Object.assign(workHours, { objectId: itemId, type: "WORKHOURS" });
  return request({
    url: `/tracker/project/items/${itemId}/workHours`,
    method: METHOD.DELETE,
    data: workHours,
  });
}
export function remainingRegistrableTime(memberId, beginDate) {
  return request({
    url: `/tracker/project/items/${memberId}/workHours/member`,
    method: METHOD.GET,
    params: beginDate,
  });
}
export function findAssociations() {
  return request({
    url: `/tracker/project/items/associations`,
    method: METHOD.GET,
  });
}
export function updateTrackerItemSprint(sprintId, trackerItemIds) {
  return request({
    url: `/tracker/project/items/sprint`,
    method: METHOD.PUT,
    params: { sprintId },
    data: trackerItemIds,
  });
}
export function findTrackersBySprint(sprintId) {
  return request({
    url: `/tracker/project/items/sprint/${sprintId}/trackers`,
    method: METHOD.GET,
  });
}

export function cardInfo(projectId) {
  return request({
    url: `/tracker/project/items/cardInfo/${projectId}`,
    method: METHOD.GET,
  });
}

export function updateTrackerItemTestStep(itemId, testStep) {
  return request({
    url: `/tracker/project/items/${itemId}/testStep`,
    method: METHOD.PUT,
    data: testStep,
  });
}
export function updateTrackerItemTestSteps(itemId, testSteps) {
  return request({
    url: `/tracker/project/items/${itemId}/testSteps`,
    method: METHOD.PUT,
    data: testSteps,
  });
}
export function findTrackerLinksByItemIds(itemIds) {
  return request({
    url:
      "/tracker/project/items/trackerLinks",
    method: METHOD.POST,
    data: itemIds
  });
}
export function updateMatrixLinks(trackerLInks) {
  return request({
    url:
      "/tracker/project/items/updateMatrixLinks",
      method: METHOD.POST,
      data: trackerLInks,
  });
}

File: server/helm-tracker/src/main/java/com/harbortek/helm/tracker/service/TrackerItemService.java
/*
 * Copyright [2025] [Harbortek Corp.]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.harbortek.helm.tracker.service;

import com.harbortek.helm.common.vo.BaseIdentity;
import com.harbortek.helm.tracker.entity.link.TrackerLinkEntity;
import com.harbortek.helm.tracker.entity.log.HyperlinkEntity;
import com.harbortek.helm.tracker.entity.log.WorkHoursEntity;
import com.harbortek.helm.tracker.entity.tracker.TrackerEntity;
import com.harbortek.helm.tracker.vo.chart.ProjectCardInfo;
import com.harbortek.helm.tracker.vo.items.TrackerItemVo;
import com.harbortek.helm.tracker.vo.link.TrackerLinkVo;
import com.harbortek.helm.tracker.vo.log.AttachmentVo;
import com.harbortek.helm.tracker.vo.log.CommentVo;
import com.harbortek.helm.tracker.vo.log.TrackerLogVo;
import com.harbortek.helm.tracker.vo.log.WorkHoursVo;
import com.harbortek.helm.tracker.vo.tracker.stateTransition.TrackerStatus;
import com.harbortek.helm.tracker.vo.view.FilterCondition;
import com.harbortek.helm.tracker.vo.view.ObjectFilter;
import com.harbortek.helm.util.excel.ExcelLogs;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.util.Date;
import java.util.List;

public interface TrackerItemService {
    Page<TrackerItemVo> findTrackerItems(Long projectId, Long trackerId, Long sprintId, ObjectFilter filter,
                                         String keyword, Boolean isTest, Pageable pageable);
    Page<TrackerItemVo> findTrackerItemsTree(Long projectId, Long trackerId, Long sprintId, ObjectFilter filter,
                                             String keyword, Boolean LinkDire, Long linkType, Integer linkLevel, Pageable pageable);

    TrackerItemVo createTrackerItem(TrackerItemVo trackerItemVo);

    void updateTrackerItem(TrackerItemVo trackerItemVo);

    TrackerItemVo findOneTrackerItem(Long itemId);

    List<TrackerItemVo> findTrackerItemByIds(List<Long> itemIds);
    List<TrackerItemVo> findByTrackerIds(Long projectId, List<Long> trackerIds);
    void deleteOneTrackerItem(Long itemId);

    void batchDeleteTrackerItem(List<Long> itemIds);

    TrackerStatus stateChanged(Long itemId, Long stateTransitionId);

    void fieldChanged(Long itemId, Long fieldId, Object newValue);

    void systemFieldChanged(Long itemId, String systemProperty, Object newValue);

    List<TrackerLogVo> findCTrackerLogs(Long itemId, Boolean includeComment, Boolean includeChangeLog, Boolean includeAttachment, Boolean includeAssociation);

    void createComment(CommentVo comment);

    void saveComment(CommentVo comment);

    void deleteComment(CommentVo comment);

    void createWorkHours(WorkHoursEntity workHours);

    void saveWorkHours(WorkHoursEntity workHours);

    void deleteWorkHours(WorkHoursEntity workHours);

    void replyComment(CommentVo comment);

    AttachmentVo uploadAttachment(Long itemId, AttachmentVo attachmentVo);

    void deleteAttachment(Long itemId, AttachmentVo attachmentVo);

    List<AttachmentVo> findAttachmentsByObjectId(Long objectId);

    void batchCreateTrackerLinks(Long itemId, List<TrackerLinkVo> trackerLink);
    void deleteTrackerLink(Long itemId, TrackerLinkVo trackerLink);
    boolean updateTrackerLink(Long itemId, TrackerLinkVo trackerLink);
    List<HyperlinkEntity> findHyperlinkByObjectId(Long objectId);
    HyperlinkEntity updateHyperlink(HyperlinkEntity hyperlink);
    void deleteHyperlink(Long hyperlinkId);

    Double remainingRegistrableTime(Long memberId, Date beginDate);

    List<WorkHoursVo> findWorkHours(Long itemId);

    void updateTrackerItemSprint(Long sprintId,List<Long> trackerItemIds);

    void relatedWikis(Long itemId, List<Long> relatedWikiIds);

    void exportTrackerItems(HttpServletRequest request, HttpServletResponse response, Long projectId, Long trackerId, Long sprintId, List<Long> selectedRowIds, ObjectFilter filter, PageRequest of);

    void downloadImportDome(HttpServletRequest request, HttpServletResponse response, Long trackerId);

    ExcelLogs importTrackerItems(Long trackerId, String uploadedFile);

    void watch(Long itemId, BaseIdentity watch);

    void cancelWatch(Long itemId, BaseIdentity watch);

    void batchUpdateTrackerItem(List<TrackerItemVo> trackerItemVos);

    List<TrackerEntity> findTrackersBySprint(Long sprintId);

    ProjectCardInfo findProjectCardInfo(Long projectId);

//    void updateTrackerItemTestStep(Long itemId, TrackerTestStep testStep);
//
//    void updateTrackerItemTestSteps(Long itemId, List<TrackerTestStep> testSteps);

    List<FilterCondition> findGroupItems(Long projectId, Long trackerId, Long sprintId, ObjectFilter filter, String keyword);

    void deleteTrackerItemsByTrackerId(Long trackerId);

    List<TrackerLinkVo> findTrackerLinksByItemIds(List<Long> itemIds);

    void updateMatrixLinks(List<TrackerLinkEntity> trackerLinks);


     void setFieldValue(TrackerItemVo trackerItem, Long fieldId, Object value);
}
